### ISSUE TITLE
JVM, .NET Language Support

### ISSUE THREAD

**Issue by sirinath** [tensorflow/tensorflow#3]
Your website says "... contribute SWIG interfaces to your favorite language ...". It will be good if yo have pure implementations in other languages like JVM and .NET without having to use SWIG.

**Comment by sheepdreamofandroids** [comment]
Since raw performance is one of the goals of this library I think that pure implementations don't make too much sense.
A prettier API on top of the generated interfaces would be nice though.

**Comment by girving** [comment]
Although we may not pursue this ourselves, the internal formats used for communication (`GraphDef`, etc.) are all platform independent protobuf.  Thus, it is possible to implement all or part of the tensorflow API in Java.  Since a Java version would likely be slower, one useful bit would be a pure inference layer that evaluates graphs but isn't necessarily able to build them; this would allow graphs built in Python and trained in Python / C++ on GPUs to be run from Java servers.

**Comment by edblackburn** [comment]
Is there a canonical / repeatable test suite, so language bindings can have a target and level of confidence? Is there a build server? Could there be a build server?

**Comment by girving** [comment]
There's a testsuite with fairly good converge, but currently it's mostly Python with a few C++ tests.  There's also a lot of functionality for building graphs that's currently Python only, in particular the automatic differentiation functionality, though that doesn't matter for evaluation of graphs in Java.  There are plans to move this functionality into the underlying C++ in future, at which point Java SWIG bindings would be more useful for creating graphs.

If someone takes up the Java SWIG challenge, we'd be happy to accept it upstream pending review, etc., at which point it would be part of our continuous testing.  The details of accepting contributions is in flux at the moment, but that will stabilize.

**Comment by sirinath** [comment]
Also perhaps https://github.com/bytedeco/javacpp can be used to generate bindings.

**Comment by kovasb** [comment]
@girving What aspect of automatic differentiation is not available from other languages? Is that just for defining new ops?

**Comment by girving** [comment]
If you search for `RegisterGradient` in the python directory, you'll see all the places we define the gradients of various ops.  The gradients are defined by Python code, so they aren't available from pure C++ as yet (we hope to change this at some point).

**Comment by kovasb** [comment]
I don't fully grasp the architecture/order of operations, but could one populate the gradients using python, while still calling into the C++ code via a bridge?

Alternatively: Run the python code once to generate the gradients for every op, serialize all those gradients to a file, and then other runtimes can read them off and register them?

**Comment by girving** [comment]
Both of those might work, but I don't know how much of a savings they are compared to doing it the right way and porting the code to C++.

**Comment by kovasb** [comment]
I fully agree, just sounding out a temporary solution so I'm not blocked (I imagine it might be a while before all that code is ported)

**Comment by sirinath** [comment]
If all the functionality is not available in the C++ core then porting will be very difficult and some cases not possible (with respect to a fully functional version). Also I opened this: https://github.com/tensorflow/tensorflow/issues/476

**Comment by ieee8023** [comment]
+1 for Java

**Comment by junwucs** [comment]
+1 for java

**Comment by FrankXia** [comment]
+1 for Java.

**Comment by stream1984** [comment]
+1 for Java

**Comment by thammegowda** [comment]
+1 for Java/Scala

**Comment by solomatov** [comment]
+1

**Comment by guotong1988** [comment]
@girving I want to contribute to tensorflow java api,
I used to implement the PHP frontend of Apache Spark alone.
Where to start?

**Comment by asimshankar** [comment]
Closing this out as a duplicate of #5 and #18.
